heat_pump_adoption_modelling/pipeline/MCS/mcs_epc_joining.py

```diff
@@ -0,0 +1,322 @@
+        If None, domestic HP data is loaded and augmented.
+    epcs : pandas.Dataframe
+        Dataframe with standardised_postcode, numeric_tokens
+        and standardised_address fields.
+        If None, EPC data is loaded and augmented.
+    save : bool
+        Whether or not to save the output.
+    drop_epc_address : bool
+        Whether or not to drop addresses from the EPC records.
+        Useful to keep for determining whether matches are sensible.
+
+    Return
+    ----------
+    merged : pandas.Dataframe
+        Dataframe containing merged MCS and EPC records."""
+
+    if dhps is None:
+        print("Preparing domestic HP data...")
+        dhps = load_domestic_hps()
+        dhps = prepare_dhps(dhps)
+
+    if epcs is None:
+        print("Preparing EPC data...")
+        epcs = load_epcs()
+        epcs = prepare_epcs(epcs)
+
+    print("Forming a matching...")
+    matching = form_matching(df1=dhps, df2=epcs)
+
+    # First ensure that all matches are above the matching parameter
+    good_matches = matching[
+        (matching["numerics"] == 1) & (matching["address_score"] >= matching_parameter)
+    ].reset_index()
+
+    # Take the indices of the rows with best address match for each MCS index
+    top_match_indices = good_matches.groupby("level_0")["address_score"].idxmax()
+
+    # Filter the matches df to just the top matches
+    # Drop duplicates in case of ties
+    top_matches = (
+        good_matches.loc[top_match_indices]
+        .drop_duplicates("level_0")
+        .drop(columns=["numerics", "address_score"])
+    )
+
+    print("Joining the data...")
+    merged = (
+        dhps.reset_index()
+        .merge(top_matches, how="left", left_on="index", right_on="level_0")
+        .merge(epcs.reset_index(), how="left", left_on="level_1", right_on="index")
+        .drop(
+            columns=[
+                "index_x",
+                "standardised_postcode_x",
+                "standardised_address_x",
+                "numeric_tokens_x",
+                "ADDRESS1",
+                "POSTTOWN",
+                "POSTCODE",
+                "index_y",
+                "standardised_postcode_y",
+                "numeric_tokens_y",
+            ]
+        )
+    )
+
+    if drop_epc_address:
+        merged = merged.drop(columns="standardised_address_y")
+    else:
+        merged = merged.rename({"standardised_address_y": "epc_address"})
+
+    if save:
+        merged.to_csv(PROJECT_DIR / merged_path)
+
+    return merged
+
+
+# ---------------------------------------------------------------------------------
+
+
+def main():
+    """Main function: Loads and joins MCS data to EPC data."""
+
+    start_time = time.time()
+
+    join_mcs_epc_data()
+
+    end_time = time.time()
+    runtime = round((end_time - start_time) / 60)
+
+    print(
+        "Loading and joining MCS and EPC data took {} minutes.\n\nOutput saved in ".format(
+            runtime
+        )
+        + merged_path
+    )
```

Comment:
Since you're already using format string, let's combine them:

```suggestion
    print(
        "Loading and joining MCS and EPC data took {} minutes.\n\nOutput saved in {}".format(
            runtime,  merged_path
        )
 
    )
```

Reply:
Nice, thanks!